rclcpp/include/rclcpp/mapped_ring_buffer.hpp

```diff
@@ -180,29 +235,44 @@ class MappedRingBuffer : public MappedRingBufferBase
    * It is up to the user to ensure the key is unique.
    * This method should not allocate memory.
    *
-   * After insertion, if a pair was replaced, then value will contain ownership
-   * of that displaced value. Otherwise it will be a nullptr.
+   * After insertion the value will be a nullptr.
+   * If a pair were replaced, its smart pointer is reset.
    *
    * \param key the key associated with the value to be stored
    * \param value the value to store, and optionally the value displaced
    */
   bool
-  push_and_replace(uint64_t key, ElemUniquePtr & value)
+  push_and_replace(uint64_t key, ConstElemSharedPtr value)
   {
     std::lock_guard<std::mutex> lock(data_mutex_);
     bool did_replace = elements_[head_].in_use;
-    elements_[head_].key = key;
-    elements_[head_].value.swap(value);
-    elements_[head_].in_use = true;
+    Element & element = elements_[head_];
+    element.key = key;
+    element.unique_value.reset();
+    element.shared_value.reset();
+    element.shared_value = value;
```

Comment:
Tracing this from `rclcpp::Publisher<T>::publish(std::shared_ptr<T> msg)` down to here, it appears we are going to now share the value reference by `element.shared_value` with the user's shared pointer. This is really dangerous in my opinion. If I implied we should do this, then I was mistaken.

The idea of letting the IPM store share pointers is still valid, but only for the case where converting a unique pointer to a shared pointer has some benefit due to needing to deliver it to multiple subscribers which take `std::shared_ptr<const MessageT>` to their callback.

So in short, I think we need to duplicate the message referenced by `value` here and create our own shared pointer to point to the new copy.

Reply:
I have already deleted the `rclcpp::Publisher<T>::publish(std::shared_ptr<T> msg)` method.
I only left the one taking `std::shared_ptr<const T>`

Reply:
I don't think `std::shared_ptr<const T>` is sufficiently safe, see:

https://gist.github.com/wjwwood/bf52d3911e7942077071a9089e3148e1

Reply:
Yes, that was exactly the doubt I have before. For the moment, I'm going to copy it and do a publish by value. We should later mark it as deprecated.